File: algorithms_on_strings/week1/_0741ec911710e017f3b8dc9c95001729_Programming-Assignment-1/trie_matching_extended/trie_matching_extended.cpp
#include <algorithm>
#include <cassert>
#include <cstdio>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int const Letters =    4;
int const NA      =   -1;

struct Node
{
	int next [Letters];
	bool patternEnd;

	Node ()
	{
		fill (next, next + Letters, NA);
		patternEnd = false;
	}
};

typedef vector<Node> trie;

int letterToIndex (char letter)
{
	switch (letter)
	{
		case 'A': return 0; break;
		case 'C': return 1; break;
		case 'G': return 2; break;
		case 'T': return 3; break;
		default: assert (false); return -1;
	}
}

trie build_trie(const vector<string> & patterns) {
  trie t;
  Node root;
  t.push_back(root);
  for(const string& p : patterns){
  	size_t current_node = 0;
  	for(size_t i = 0; i < p.length(); ++i){
  		size_t next_i = letterToIndex(p[i]);
  		if(t[current_node].next[next_i] != NA)
  			current_node = t[current_node].next[next_i];
  		else{
  			Node tmp;
  			t.push_back(tmp);
  			t[current_node].next[next_i] = t.size()-1;
  			current_node = t[current_node].next[next_i];
  		}
  	}
  	t[current_node].patternEnd = true;
  }
  return t;
}

vector <int> solve (const string& text, int n, const vector <string>& patterns)
{
	vector <int> result;
  trie t = build_trie(patterns);

	// write your code here
	for(size_t i = 0; i < text.length(); ++i){
	size_t current_node = 0;
	size_t current_symbol = i;
	while(true){
		if(t[current_node].patternEnd){
			result.push_back(i);
			break;
		}
		else if(current_symbol < text.length() and t[current_node].next[letterToIndex(text[current_symbol])] != NA){
			current_node = t[current_node].next[letterToIndex(text[current_symbol])];
			current_symbol++;}
		else
			break;
	}
}

	return result;
}

int main (void)
{
	string text;
	cin >> text;

	int n;
	cin >> n;

	vector <string> patterns (n);
	for (int i = 0; i < n; i++)
	{
		cin >> patterns[i];
	}

	vector <int> ans;
	ans = solve (text, n, patterns);

	for (int i = 0; i < (int) ans.size (); i++)
	{
		cout << ans[i];
		if (i + 1 < (int) ans.size ())
		{
			cout << " ";
		}
		else
		{
			cout << endl;
		}
	}

	return 0;
}

File: algorithms_on_strings/week1/_0741ec911710e017f3b8dc9c95001729_Programming-Assignment-1/trie_matching_extended/trie_matching_extended.py
# python3
import sys

NA = -1

class Node:
	#def __init__ (self):
		#self.next = [NA] * 4
		#self.patternEnd = False
    def __init__(self, pattern=None, children=None):
        if pattern is None:
            pattern = False

        if children is None:
            children = dict()

        self.pattern = pattern
        self.children = children


def outgoing_edge(tree, node_id, symbol):
    node = tree.get(node_id)
    if node is not None:
        return node.children.get(symbol)
    return None


def build_trie(patterns):
    tree = dict()
    # write your code here
    tree[0] = Node()
    max_node = 0
    for pattern in patterns:
        current_node = 0
        for idx in range(len(pattern)):
            current_symbol = pattern[idx]
            edge = outgoing_edge(tree, current_node, current_symbol)
            if edge is not None:
                current_node = tree[current_node].children[current_symbol]
            else:
                node = tree.get(current_node)
                if node is not None:
                    node.children[current_symbol] = max_node + 1
                else:
                    tree[current_node] = Node()
                    tree[current_node].children[current_symbol] = max_node + 1
                current_node = max_node + 1
                max_node = current_node

        if tree.get(current_node) is None:
            tree[current_node] = Node(pattern=True)
        else:
            tree[current_node].pattern = True

    return tree

def is_leaf(current_node, tree):
    node = tree[current_node]
    return node.pattern and (len(node.children) is 0)


def find_edge(tree, current_node, symbol):
    node = tree.get(current_node)
    if node is not None:
        for child in node.children:
            if symbol is child:
                return node.children[symbol]
    return None


def is_pattern(current_node, tree):
    node = tree.get(current_node)
    if node:
        return node.pattern
    return False

def prefix_trie_matching(text, tree):
    symbol_pos = 0
    current_node = 0
    pat_found = False
    while symbol_pos < len(text):
        symbol = text[symbol_pos]
        edge = find_edge(tree, current_node, symbol)
        if edge is not None:
            current_node = edge
            if tree[current_node].pattern:
                break
        else:
            break
        symbol_pos += 1

    if is_leaf(current_node, tree):
        pat_found = True
    elif is_pattern(current_node, tree):
        pat_found = True

    return pat_found

def solve (text, n, patterns):
    tree = build_trie(patterns)

    result = []

    txt = text

    offset = 0

    while len(txt) > 0:
        pat_found = prefix_trie_matching(txt, tree)
        if pat_found is True:
            result.append(offset)

        txt = txt[1:]
        offset += 1

	return result

text = sys.stdin.readline ().strip ()
n = int (sys.stdin.readline ().strip ())
patterns = []
for i in range (n):
	patterns += [sys.stdin.readline ().strip ()]

ans = solve (text, n, patterns)

sys.stdout.write (' '.join (map (str, ans)) + '\n')
